govc: Command to print device/backings model as tree

akutz opened the pull request:
## Description

This patch introduces the `govc device.model.tree [-backings]` command for printing either the VirtualDevice or VirtualDeviceBackingInfo type hierarchy as a tree.

Closes: `NA`

## Type of change

Please mark options that are relevant:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Build related change

## How Has This Been Tested?

1. Built `govc` with:

    ```shell
    make -C govc install
    ```

2. Ran `govc device.model.tree`:

    ```shell
    VirtualDevice
    ├── VirtualCdrom
    ├── VirtualController
    │   ├── VirtualIDEController
    │   ├── VirtualNVDIMMController
    │   ├── VirtualNVMEController
    │   ├── VirtualPCIController
    │   ├── VirtualPS2Controller
    │   ├── VirtualSATAController
    │   │   └── VirtualAHCIController
    │   ├── VirtualSCSIController
    │   │   ├── ParaVirtualSCSIController
    │   │   ├── VirtualBusLogicController
    │   │   ├── VirtualLsiLogicController
    │   │   └── VirtualLsiLogicSASController
    │   ├── VirtualSIOController
    │   ├── VirtualUSBController
    │   └── VirtualUSBXHCIController
    ├── VirtualDisk
    ├── VirtualEthernetCard
    │   ├── VirtualE1000
    │   ├── VirtualE1000e
    │   ├── VirtualPCNet32
    │   ├── VirtualSriovEthernetCard
    │   └── VirtualVmxnet
    │       ├── VirtualVmxnet2
    │       └── VirtualVmxnet3
    │           └── VirtualVmxnet3Vrdma
    ├── VirtualFloppy
    ├── VirtualKeyboard
    ├── VirtualMachineVMCIDevice
    ├── VirtualMachineVMIROM
    ├── VirtualMachineVideoCard
    ├── VirtualNVDIMM
    ├── VirtualPCIPassthrough
    ├── VirtualParallelPort
    ├── VirtualPointingDevice
    ├── VirtualPrecisionClock
    ├── VirtualSCSIPassthrough
    ├── VirtualSerialPort
    ├── VirtualSoundCard
    │   ├── VirtualEnsoniq1371
    │   ├── VirtualHdAudioCard
    │   └── VirtualSoundBlaster16
    ├── VirtualTPM
    ├── VirtualUSB
    └── VirtualWDT
    ```

3. Ran `govc device.model.tree -backings`:

    ```shell
    VirtualDeviceBackingInfo
    ├── VirtualDeviceDeviceBackingInfo
    │   ├── VirtualCdromAtapiBackingInfo
    │   ├── VirtualCdromPassthroughBackingInfo
    │   ├── VirtualDiskRawDiskVer2BackingInfo
    │   │   └── VirtualDiskPartitionedRawDiskVer2BackingInfo
    │   ├── VirtualEthernetCardLegacyNetworkBackingInfo
    │   ├── VirtualEthernetCardNetworkBackingInfo
    │   ├── VirtualFloppyDeviceBackingInfo
    │   ├── VirtualPCIPassthroughDeviceBackingInfo
    │   ├── VirtualPCIPassthroughDynamicBackingInfo
    │   ├── VirtualParallelPortDeviceBackingInfo
    │   ├── VirtualPointingDeviceDeviceBackingInfo
    │   ├── VirtualSCSIPassthroughDeviceBackingInfo
    │   ├── VirtualSerialPortDeviceBackingInfo
    │   ├── VirtualSoundCardDeviceBackingInfo
    │   ├── VirtualUSBRemoteHostBackingInfo
    │   └── VirtualUSBUSBBackingInfo
    ├── VirtualDeviceFileBackingInfo
    │   ├── VirtualCdromIsoBackingInfo
    │   ├── VirtualDiskFlatVer1BackingInfo
    │   ├── VirtualDiskFlatVer2BackingInfo
    │   ├── VirtualDiskLocalPMemBackingInfo
    │   ├── VirtualDiskRawDiskMappingVer1BackingInfo
    │   ├── VirtualDiskSeSparseBackingInfo
    │   ├── VirtualDiskSparseVer1BackingInfo
    │   ├── VirtualDiskSparseVer2BackingInfo
    │   ├── VirtualFloppyImageBackingInfo
    │   ├── VirtualNVDIMMBackingInfo
    │   ├── VirtualParallelPortFileBackingInfo
    │   └── VirtualSerialPortFileBackingInfo
    ├── VirtualDevicePipeBackingInfo
    │   └── VirtualSerialPortPipeBackingInfo
    ├── VirtualDeviceRemoteDeviceBackingInfo
    │   ├── VirtualCdromRemoteAtapiBackingInfo
    │   ├── VirtualCdromRemotePassthroughBackingInfo
    │   ├── VirtualFloppyRemoteDeviceBackingInfo
    │   └── VirtualUSBRemoteClientBackingInfo
    ├── VirtualDeviceURIBackingInfo
    │   └── VirtualSerialPortURIBackingInfo
    ├── VirtualEthernetCardDistributedVirtualPortBackingInfo
    ├── VirtualEthernetCardOpaqueNetworkBackingInfo
    ├── VirtualPCIPassthroughDvxBackingInfo
    ├── VirtualPCIPassthroughPluginBackingInfo
    │   └── VirtualPCIPassthroughVmiopBackingInfo
    ├── VirtualPrecisionClockSystemClockBackingInfo
    ├── VirtualSerialPortThinPrintBackingInfo
    └── VirtualSriovEthernetCardSriovBackingInfo
    ```

## Checklist:

- [x] My code follows the `CONTRIBUTION` [guidelines] of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

[guidelines]: https://github.com/vmware/govmomi/blob/main/CONTRIBUTING.md

dougm reviewed (commented):
Very cool, thank you @akutz !  I still haven't touched Go generics, so I'll need to read guide, then read this a few more times for my own education :)

Only issue is the the linter failed, but not with a very helpful error message. Can see it locally:

```console
% goimports -d -local github.com/vmware/govmomi govc/device/model/tree.go 
diff -u govc/device/model/tree.go.orig govc/device/model/tree.go
--- govc/device/model/tree.go.orig	2023-09-14 17:10:48
+++ govc/device/model/tree.go	2023-09-14 17:10:48
@@ -24,9 +24,10 @@
 	"sort"
 	"unsafe"
 
+	"github.com/xlab/treeprint"
+
 	"github.com/vmware/govmomi/govc/cli"
 	"github.com/vmware/govmomi/vim25/types"
-	"github.com/xlab/treeprint"
 )
 
 type tree struct {
```

akutz commented:
Thanks @dougm , I addressed the issue. I also added the ability to search for a given type and print just that type's tree:

```shell
$ govc device.model.tree VirtualEthernetCard                        
VirtualEthernetCard
├── VirtualE1000
├── VirtualE1000e
├── VirtualPCNet32
├── VirtualSriovEthernetCard
└── VirtualVmxnet
    ├── VirtualVmxnet2
    └── VirtualVmxnet3
        └── VirtualVmxnet3Vrdma
```

```shell
$ govc device.model.tree -backings VirtualDiskRawDiskVer2BackingInfo
VirtualDiskRawDiskVer2BackingInfo
└── VirtualDiskPartitionedRawDiskVer2BackingInfo
```

dougm reviewed (approved):
lgtm
